broadcast.go

```diff
@@ -35,21 +102,32 @@ type MessageStore struct {
 }
 
 func (s *MessageStore) Initialize(n *maelstrom.Node, topology Topology) {
+	s.node = n
 	s.replicationQueue = startReplicationWorkers(n, topology[n.ID()])
 }
 
 func (s *MessageStore) Write(message BroadcastMessage) bool {
 	s.mutex.Lock()
-	defer s.mutex.Unlock()
 
 	_, ok := s.hashMap[message.Body.Message]
 
 	if !ok {
 		s.hashMap[message.Body.Message] = true
 		s.slice = append(s.slice, message.Body.Message)
+		storeSize.WithLabelValues(s.node.ID()).Inc()
+	}
+
+	s.mutex.Unlock()
+
+	// Inter-server messages don't need to be replicated
+	if strings.HasPrefix(message.Src, "c") {
```

Comment:
This was the bug that took a long time to find. 

We re-use the `broadcast` RPC to replicate messages between nodes. However, we only need to replicate messages that we receive from a client - if we replicate messages received from other nodes, we end up with infinite loops.

I had been checking `msg_id` to determine if the message came from a client or node, as described in the [first section of the old tutorial](https://fly.io/dist-sys/3d/), but was getting the aforementioned infinite loops. I read the source of how RPC messages are sent in the golang client and discovered it adds `msg_id`. Checking the prefix of the `src` is crude - clients have IDs like `c1` and nodes have IDs like `n1` - but it works.